src/shared/openssl-util.c

```diff
@@ -1130,6 +1129,101 @@ int string_hashsum(
 }
 #  endif
 
+static int ecc_pkey_generate_volume_key(
+                EVP_PKEY *pkey,
+                void **ret_decrypted_key,
+                size_t *ret_decrypted_key_size,
+                void **ret_saved_key,
+                size_t *ret_saved_key_size) {
+
+        _cleanup_(EVP_PKEY_freep) EVP_PKEY *pkey_new = NULL;
+        _cleanup_(erase_and_freep) void *decrypted_key = NULL;
+        _cleanup_free_ unsigned char *saved_key = NULL;
+        size_t decrypted_key_size, saved_key_size;
+        int nid = NID_undef;
+        int r;
+
+#if OPENSSL_VERSION_MAJOR >= 3
+        _cleanup_free_ char *curve_name = NULL;
+        size_t len = 0;
+
+        if (EVP_PKEY_get_group_name(pkey, NULL, 0, &len) != 1 || len == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to determine PKEY group name length");
```

Comment:
use log_openssl_errors here?

Reply:
log_openssl_errors() logs at debug level, but we need a error level here.

Reply:
actually, I'd say this log message here should probably be at debug level.

in systemd we have this rough rule that "library like" code should log at most at debug level, and "main function" code should log above debug logging. What kind of code each function is not always clear cut, there's a lot of blur in between. But usually generic utility functions like this one should only debug log.

See: https://systemd.io/CODING_STYLE/#error-handling